### PR TITLE
Enabling user-specified scorer in HyperoptRegressor + best score

### PR DESCRIPTION
**Pull request by rithram**
This PR contains 2 changes:
- We allow the user to set the `best_score` in which case the `loss` is computed as `best_score - cv_score`. This makes the `loss >= 0` is `best_score` is correctly set. I am not entirely sure if this affect hyperopt at all but there are many black-box optimizers that perform better with positive objectives that they are trying to minimize rather than negative objectives that is being minimized. If the user does not specify the `best_score`, current behavior (which is equivalent to `best_score = 0` is preserved.
- Update `HyperoptRegressor` to allow user to specify the scorer similar to `HyperoptClassifier`.

### CODE REVIEW

**Comment by rithram** [comment]
@kiran-kate @hirzel 
So I updated the schema for `HyperoptRegressor` and now it is a `TrainableOperator`. However, the return of the `.fit` call is not a `Trained` `HyperoptRegressor` and hence does not have the `get_trials()` and related functionality, and currently there is no way to access the `get_trials()` for this operator.

File: lale/lib/lale/hyperopt_regressor.py [hunk 1]
```diff
@@ -86,14 +88,17 @@ def f(params):
...
+                    'loss': self.best_score - score, 'time': execution_time, 'status': STATUS_OK
+                }
             except BaseException as e:
-                logger.warning(f'Exception caught in HyperoptRegressor: {type(e)}, {traceback.format_exc()} with hyperparams: {params}, setting loss to zero')
-                r_squared = 0
-                execution_time = 0
-            return {'loss': -r_squared, 'time': execution_time, 'status': STATUS_OK}
 
+                logger.warning(f'Exception caught in HyperoptRegressor: {type(e)}, {traceback.format_exc()} with hyperparams: {params}, setting loss to zero')
+                return_dict = {'status': STATUS_FAIL}
```
Comment (rithram) [inline lale/lib/lale/hyperopt_regressor.py]: I have changed this to return a `STATUS_FAIL` since I could not think of a good default loss function for `r2_score` (it can be arbitrarily high).

**Comment by hirzel** [comment]
Hi Pari, Thanks a lot for implementing the scoring for `HyperoptRegressor`! We often get asked how to use other metrics, so it's good to have a clean answer for that question. So I'll merge your PR.

Regarding your comment about the `TrainableOperator`, sorry, this was caused by some other independent changes I had made. Thanks for pointing out that that temporarily hides `get_trials()`. Let's get that functionality back using a separate PR.
